Display can go to sleep when camera is on

youennf opened the pull request:
#### 79a3c4d8d291353d877baa33fd41bb4c104e2abc
<pre>
Display can go to sleep when camera is on
<a href="https://bugs.webkit.org/show_bug.cgi?id=245726">https://bugs.webkit.org/show_bug.cgi?id=245726</a>
rdar://100423979

Reviewed by Eric Carlson.

In case a document is capturing or screen, it is good to keep the display on so that the preview stays visible.
To do so, the document creates a SleepDisabler when capturing.

Covered by updated API tests.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::mediaStreamCaptureStateChanged):
(WebCore::Document::canvasChanged):
(WebCore::Document::updateSleepDisablerIfNeeded):
* Source/WebCore/dom/Document.h:
(WebCore::Document::hasSleepDisabler const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::hasSleepDisabler const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Tools/TestWebKitAPI/Tests/WebKit/GetUserMedia.mm:
* Tools/TestWebKitAPI/Tests/WebKit/SleepDisabler.mm:
(TEST_F):
* Tools/TestWebKitAPI/Tests/WebKit/getDisplayMedia.html:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/GetDisplayMediaWindowAndScreen.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitLegacy/ios/video-with-audio.html:

Canonical link: <a href="https://commits.webkit.org/255636@main">https://commits.webkit.org/255636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ad7cd31870502097961db8d477908ddbcf578e4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93092 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2292 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23735 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102799 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163107 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97094 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2300 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30620 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85482 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98917 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98757 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1585 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79566 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28505 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83499 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83262 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71614 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37013 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17148 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34825 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18381 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3905 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38696 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40923 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40626 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37585 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9150,6 +9159,29 @@ void Document::canvasChanged(CanvasBase& canvasBase, const std::optional<FloatRe
+
+    m_sleepDisablerTimer.startRepeating(5_s);
+    m_sleepDisabler = makeUnique<SleepDisabler>("com.apple.WebCore: Document playing audio with active canvas"_s, PAL::SleepDisabler::Type::Display);
+    m_canvasChangedWhileDisablingSleep = false;
```
Nit: I think this would be slightly clearer if you set `m_canvasChangedWhileDisablingSleep = true` before checking to see if the timer is active, and don't set `m_canvasChangedWhileDisablingSleep = false` at all in this method.

eric-carlson reviewed (approved):
r=me

jernoble reviewed (changes requested):
It seems much, much too easy to accidentally trigger this heuristic on sites that use AudioContext and canvas for _jon_ video conferencing reasons.

eric-carlson commented on Tools/TestWebKitAPI/Tests/WebKit/SleepDisabler.mm:
```diff
@@ -245,3 +245,22 @@ void hasSleepDisablerShouldBecomeEqualTo(bool shouldHaveSleepDisabler)
+    hasSleepDisablerShouldBecomeEqualTo(false);
+
+    [webView evaluateJavaScript:@"startCanvasUpdate()" completionHandler:nil];
+    hasSleepDisablerShouldBecomeEqualTo(true);
```
Is this test still valid?

youennf replied:
Right, I will remove it.

eric-carlson commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9171,6 +9174,16 @@ void Document::canvasChanged(CanvasBase& canvasBase, const std::optional<FloatRe
+    MediaProducerMediaStateFlags activeVideoCaptureMask { MediaProducerMediaState::HasActiveVideoCaptureDevice, MediaProducerMediaState::HasActiveScreenCaptureDevice, MediaProducerMediaState::HasActiveWindowCaptureDevice };
+    if (m_mediaState & activeVideoCaptureMask) {
+        m_sleepDisabler = makeUnique<SleepDisabler>("com.apple.WebCore: Document doing camera, screen or window capture"_s, PAL::SleepDisabler::Type::Display);
+        return;
```
Could this return early if `m_sleepDisabler` is non-null instead of recreating it?

youennf replied:
Good idea, will do

eric-carlson reviewed (commented):
The commit message should be updated for the new strategy, it still talks about preventing sleep when canvas is active.

youennf committed:
Display can go to sleep when camera is on
https://bugs.webkit.org/show_bug.cgi?id=245726
rdar://100423979

Reviewed by Eric Carlson.

In case a document is capturing or screen, it is good to keep the display on so that the preview stays visible.
To do so, the document creates a SleepDisabler when capturing.

Covered by updated API tests.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::mediaStreamCaptureStateChanged):
(WebCore::Document::canvasChanged):
(WebCore::Document::updateSleepDisablerIfNeeded):
* Source/WebCore/dom/Document.h:
(WebCore::Document::hasSleepDisabler const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::hasSleepDisabler const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Tools/TestWebKitAPI/Tests/WebKit/GetUserMedia.mm:
* Tools/TestWebKitAPI/Tests/WebKit/SleepDisabler.mm:
(TEST_F):
* Tools/TestWebKitAPI/Tests/WebKit/getDisplayMedia.html:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/GetDisplayMediaWindowAndScreen.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitLegacy/ios/video-with-audio.html:

Canonical link: https://commits.webkit.org/255636@main
